src/js/tech/ChromecastTech.js

```diff
@@ -0,0 +1,364 @@
+'use strict';
+
+var ChromecastSessionManager = require('../chromecast/ChromecastSessionManager'),
+    ChromecastTechUI = require('./ChromecastTechUI'),
+    _ = require('underscore'),
+    SESSION_TIMEOUT = 10 * 1000, // milliseconds
+    ChromecastTech;
+
+
+ChromecastTech = {
+   constructor: function(options) {
+      var subclass;
+
+      this._eventListeners = [];
+
+      // We have to initialize the UI here, before calling super.constructor
+      // because the constructor calls `createEl`, which references `this._ui`.
+      this._ui = new ChromecastTechUI();
+      this._ui.updatePoster(options.source.poster);
+
+      // Call the super class' constructor function
+      subclass = this.constructor.super_.apply(this, arguments);
+
+      this._chromecastSessionManager = options.source.chromecastSessionManager;
+      this._remotePlayer = this._chromecastSessionManager.getRemotePlayer();
+      this._remotePlayerController = this._chromecastSessionManager.getRemotePlayerController();
+      this._listenToPlayerControllerEvents();
+      this.on('dispose', this._removeAllEventListeners.bind(this));
+
+      this._hasPlayedAnyItem = false;
+      this._currentSource = options.source;
+      this._requestTitle = options.requestTitleFn || _.noop;
+      this._requestSubtitle = options.requestSubtitleFn || _.noop;
+      // See `currentTime` function
+      this._firstItemStartTime = options.startTime || 0;
+      // See `duration` function
+      this._firstItemDuration = options.source.duration || 0;
+
+      this._playSource(options.source, this._firstItemStartTime);
+      this.ready(function() {
+         this.setMuted(options.muted);
+      }.bind(this));
+
+      return subclass;
+   },
+
+   createEl: function() {
+      return this._ui.getDOMElement();
+   },
+
+   play: function() {
+      if (!this.paused()) {
+         return;
+      }
+      if (this.ended() && !this._isMediaLoading) {
+         // Restart the current item from the beginning
+         this._playSource(this._currentSource, 0);
+      } else {
+         this._remotePlayerController.playOrPause();
+      }
+   },
+
+   pause: function() {
+      if (!this.paused() && this._remotePlayer.canPause) {
+         this._remotePlayerController.playOrPause();
+      }
+   },
+
+   paused: function() {
+      return this._remotePlayer.isPaused || this.ended();
+   },
+
+   setSource: function(source) {
+      // Store the current source so that we have a reference to it if we need to restart
+      // the current item.
+      this._currentSource = source;
```

Comment:
Yes, it's needed. The `_currentSource` is used when restarting the current item from the beginning (see the `play` function). The Tech implementation does not actually have a reference to the `player` (see: http://docs.videojs.com/component.js.html line 55), so calling `currentSource` is self-referential and thus the need to store the `currentSource`. If you see a better way, let me know.